api/applications.go

```diff
@@ -0,0 +1,177 @@
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+)
+
+// App represents an application deployed on Section
+type App struct {
+	ID              int           `json:"id"`
+	Href            string        `json:"href"`
+	ApplicationName string        `json:"application_name"`
+	Environments    []Environment `json:"environments"`
+}
+
+// Environment represents an application's environments on Section
+type Environment struct {
+	ID              int      `json:"id"`
+	Href            string   `json:"href"`
+	EnvironmentName string   `json:"environment_name"`
+	Domains         []Domain `json:"domains"`
+	Stack           []Proxy  `json:"stack"`
+}
+
+// Domain represents an applications environments' domains
+type Domain struct {
+	Name     string `json:"name"`
+	ZoneName string `json:"zoneName"`
+	CNAME    string `json:"cname"`
+	Mode     string `json:"mode"`
+}
+
+// Proxy represents a proxy in the traffic delivery stack
+type Proxy struct {
+	Name  string `json:"name"`
+	Image string `json:"image"`
+	Href  string `json:"href"`
+}
+
+// Application returns detailed information about a given application.
+func Application(accountID int, applicationID int) (a App, err error) {
+	u, err := url.Parse(BaseURL)
+	if err != nil {
+		return a, err
+	}
+	u.Path += fmt.Sprintf("/account/%d/application/%d", accountID, applicationID)
+
+	resp, err := request(http.MethodGet, u.String(), nil)
+	if err != nil {
+		return a, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return a, fmt.Errorf("request failed with status %s", resp.Status)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return a, err
+	}
+
+	err = json.Unmarshal(body, &a)
+	if err != nil {
+		return a, err
+	}
+
+	envs, err := ApplicationEnvironments(accountID, applicationID)
+	if err != nil {
+		return a, err
+	}
+	a.Environments = envs
+
+	for i, e := range a.Environments {
+		stack, err := ApplicationEnvironmentStack(accountID, applicationID, e.EnvironmentName)
```

Comment:
Would be nice to better understand some use cases if we can provide a native API server side to avoid the "N+1 query" problem.

Reply:
💯 

I've been writing down where I'm finding the API lacking as I go. 

This particular part of the API is a bit cludgy.

We should definitely go back and refine some of the existing API endpoints based on what I'm learning here.

Reply:
You should not touch the existing API endpoinnts. There is a new GraphQL schema. You can add them there.

Reply:
We should avoid using the old APIs from now on and only use the GraphQL API if we want to decomission Sails.

Reply:
Good call. 

We'll need to come back and rework the API client to work with the new GraphQL API after launch.